File: reference/src/telemetry/logging.ts
/**
 * Structured logging for AI-SDLC Framework (PRD Section 14).
 * Provides structured log entries with level, message, attributes, and timestamps.
 */

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

export interface LogEntry {
  level: LogLevel;
  message: string;
  timestamp: string;
  logger?: string;
  attributes?: Record<string, unknown>;
  error?: string;
}

export interface StructuredLogger {
  debug(msg: string, attrs?: Record<string, unknown>): void;
  info(msg: string, attrs?: Record<string, unknown>): void;
  warn(msg: string, attrs?: Record<string, unknown>): void;
  error(msg: string, attrs?: Record<string, unknown>, err?: Error): void;
}

export interface BufferLogger extends StructuredLogger {
  getEntries(): readonly LogEntry[];
  clear(): void;
}

function createEntry(
  level: LogLevel,
  message: string,
  logger?: string,
  attrs?: Record<string, unknown>,
  err?: Error,
): LogEntry {
  return {
    level,
    message,
    timestamp: new Date().toISOString(),
    logger,
    attributes: attrs,
    error: err?.message,
  };
}

/**
 * Create a no-op logger that silently discards all messages.
 */
export function createNoOpLogger(): StructuredLogger {
  return {
    debug() {},
    info() {},
    warn() {},
    error() {},
  };
}

/**
 * Create a buffer logger that stores entries in-memory for testing.
 */
export function createBufferLogger(name?: string): BufferLogger {
  const entries: LogEntry[] = [];

  return {
    debug(msg: string, attrs?: Record<string, unknown>) {
      entries.push(createEntry('debug', msg, name, attrs));
    },
    info(msg: string, attrs?: Record<string, unknown>) {
      entries.push(createEntry('info', msg, name, attrs));
    },
    warn(msg: string, attrs?: Record<string, unknown>) {
      entries.push(createEntry('warn', msg, name, attrs));
    },
    error(msg: string, attrs?: Record<string, unknown>, err?: Error) {
      entries.push(createEntry('error', msg, name, attrs, err));
    },
    getEntries(): readonly LogEntry[] {
      return entries;
    },
    clear() {
      entries.length = 0;
    },
  };
}

/**
 * Create a console logger that writes JSON-formatted structured logs.
 */
export function createConsoleLogger(name?: string): StructuredLogger {
  function write(level: LogLevel, msg: string, attrs?: Record<string, unknown>, err?: Error) {
    const entry = createEntry(level, msg, name, attrs, err);
    const json = JSON.stringify(entry);
    switch (level) {
      case 'debug':
        console.debug(json);
        break;
      case 'info':
        console.log(json);
        break;
      case 'warn':
        console.warn(json);
        break;
      case 'error':
        console.error(json);
        break;
    }
  }

  return {
    debug(msg: string, attrs?: Record<string, unknown>) {
      write('debug', msg, attrs);
    },
    info(msg: string, attrs?: Record<string, unknown>) {
      write('info', msg, attrs);
    },
    warn(msg: string, attrs?: Record<string, unknown>) {
      write('warn', msg, attrs);
    },
    error(msg: string, attrs?: Record<string, unknown>, err?: Error) {
      write('error', msg, attrs, err);
    },
  };
}

File: sdk-go/telemetry/logging.go
package telemetry

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sync"
	"time"
)

// LogLevel represents logging severity.
type LogLevel string

const (
	LogDebug LogLevel = "debug"
	LogInfo  LogLevel = "info"
	LogWarn  LogLevel = "warn"
	LogError LogLevel = "error"
)

// LogEntry represents a structured log entry.
type LogEntry struct {
	Timestamp string                 `json:"timestamp"`
	Level     LogLevel               `json:"level"`
	Message   string                 `json:"message"`
	Fields    map[string]interface{} `json:"fields,omitempty"`
}

// StructuredLogger defines the interface for structured logging.
type StructuredLogger interface {
	Debug(msg string, fields map[string]interface{})
	Info(msg string, fields map[string]interface{})
	Warn(msg string, fields map[string]interface{})
	Error(msg string, fields map[string]interface{})
}

// NoOpLogger discards all log messages.
type NoOpLogger struct{}

func NewNoOpLogger() *NoOpLogger { return &NoOpLogger{} }

func (l *NoOpLogger) Debug(msg string, fields map[string]interface{}) {}
func (l *NoOpLogger) Info(msg string, fields map[string]interface{})  {}
func (l *NoOpLogger) Warn(msg string, fields map[string]interface{})  {}
func (l *NoOpLogger) Error(msg string, fields map[string]interface{}) {}

// BufferLogger stores log entries in memory for inspection.
type BufferLogger struct {
	mu      sync.Mutex
	Entries []LogEntry
}

func NewBufferLogger() *BufferLogger {
	return &BufferLogger{}
}

func (l *BufferLogger) log(level LogLevel, msg string, fields map[string]interface{}) {
	l.mu.Lock()
	defer l.mu.Unlock()
	l.Entries = append(l.Entries, LogEntry{
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Level:     level,
		Message:   msg,
		Fields:    fields,
	})
}

func (l *BufferLogger) Debug(msg string, fields map[string]interface{}) { l.log(LogDebug, msg, fields) }
func (l *BufferLogger) Info(msg string, fields map[string]interface{})  { l.log(LogInfo, msg, fields) }
func (l *BufferLogger) Warn(msg string, fields map[string]interface{})  { l.log(LogWarn, msg, fields) }
func (l *BufferLogger) Error(msg string, fields map[string]interface{}) { l.log(LogError, msg, fields) }

// ConsoleLogger writes JSON log entries to an io.Writer.
type ConsoleLogger struct {
	writer io.Writer
	mu     sync.Mutex
}

func NewConsoleLogger() *ConsoleLogger {
	return &ConsoleLogger{writer: os.Stderr}
}

func NewConsoleLoggerWithWriter(w io.Writer) *ConsoleLogger {
	return &ConsoleLogger{writer: w}
}

func (l *ConsoleLogger) log(level LogLevel, msg string, fields map[string]interface{}) {
	entry := LogEntry{
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Level:     level,
		Message:   msg,
		Fields:    fields,
	}
	data, _ := json.Marshal(entry)
	l.mu.Lock()
	defer l.mu.Unlock()
	fmt.Fprintln(l.writer, string(data))
}

func (l *ConsoleLogger) Debug(msg string, fields map[string]interface{}) { l.log(LogDebug, msg, fields) }
func (l *ConsoleLogger) Info(msg string, fields map[string]interface{})  { l.log(LogInfo, msg, fields) }
func (l *ConsoleLogger) Warn(msg string, fields map[string]interface{})  { l.log(LogWarn, msg, fields) }
func (l *ConsoleLogger) Error(msg string, fields map[string]interface{}) { l.log(LogError, msg, fields) }
